Kirill_Suptelov_Task_1_Junior

TheCoolerSuptelov opened the pull request.

TheCoolerSuptelov committed:
dev

TheCoolerSuptelov committed:
Update README.md

AndryLat commented on pom.xml:
```diff
@@ -0,0 +1,74 @@
+		<version>2.7.0</version>
+		<relativePath/> <!-- lookup parent from repository -->
+	</parent>
+	<groupId>com.github.theCoolerSuptelov.redisLikeInMemory</groupId>
```
redisLikeInMemory тут не надо:)

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/CacheUnit/CacheUnit.java:
```diff
@@ -0,0 +1,56 @@
+public class CacheUnit implements Serializable {
+    private Long expiresAtMilliseconds;
+    private Long timeToLive = 60l;
+    private TimeUnit timeUnitTimeToLive = TimeUnit.SECONDS;
```
Почему не вынест ив константы?

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/CacheUnit/CachedValueDTO.java:
```diff
@@ -0,0 +1,36 @@
+package com.github.theCoolerSuptelov.redisLikeInMemory.redisLikeSpringApp.CacheUnit;
+
+import java.util.Collection;
+
```
По хорошему классы организоываем так:
сначала константы, от публичных к приватным,
потому поля класса
потом конструкторы
потом методы от публичной бизнес логики к приватным сугубо техническим методам.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/CacheUnit/CachedValueDTO.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public Long getExpiresAfterSeconds() {
+        return expiresAtMilliseconds;
```
<img width="319" alt="image" src="https://user-images.githubusercontent.com/2914310/173603131-c128e432-927e-4f9c-a5cb-6df503fea18c.png">
 меня как-то смутило, что в методе секунды, а тут миллисекунды...

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Configs/SpringSecurityConfig.java:
```diff
@@ -0,0 +1,48 @@
+public class SpringSecurityConfig {
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        // @formatter:off
```
это не надо использовать в коде.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Configs/SpringSecurityConfig.java:
```diff
@@ -0,0 +1,48 @@
+                .build();
+
+        return new InMemoryUserDetailsManager(user,userAdmin);
+    }}
```
форматирование не ок.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Controllers/RedisLikeController.java:
```diff
@@ -0,0 +1,89 @@
+
+
+@RestController
+@RequestMapping("/api")
```
Странный эндпоинт - api. Я бы рутовым эндпоинтом сделал cache

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Controllers/RedisLikeController.java:
```diff
@@ -0,0 +1,89 @@
+    private final CacheImplService cacheImplService;
+    private final String defaultAnswer = "OK";
+
+    public RedisLikeController(@Autowired CacheImplService cacheImplService) {
```
в конструкторе автовайред не надо.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Controllers/RedisLikeController.java:
```diff
@@ -0,0 +1,89 @@
+    public CachedValueDTO getValue(@RequestParam("key") String key) {
+        CacheUnit currentCacheUnit = cacheImplService. getCachingValue(key);
+        if (currentCacheUnit == null) {
+            return null;
```
эм... а что такое нал? куда мы это приплетаем? Если мы не смогли получить, то возвращаем 404 код, типа ничего не найдено, как минимум.

TheCoolerSuptelov replied:
Мне тоже этот момент показался странным, ибо строить логики поведения от null -  не красиво. Но решил сделать точно как описано в документации https://redis.io/commands/get/

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Controllers/RedisLikeController.java:
```diff
@@ -0,0 +1,89 @@
+
+    @GetMapping
+    public CachedValueDTO getValue(@RequestParam("key") String key) {
+        CacheUnit currentCacheUnit = cacheImplService. getCachingValue(key);
```
Зачем пробел после точки?:)

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Controllers/RedisLikeController.java:
```diff
@@ -0,0 +1,89 @@
+            return null;
+        }
+        if (putAnswer.equals("")){
+            return defaultAnswer;
```
просто юзеру присылаем текст ок? Не надо так. если нечего слать - шли пустоту и корректный код.

TheCoolerSuptelov replied:
Учел. Сделаю DTO ответа. Value для совместимостью с документацией оставлю как в документации, плюс положу полезную нагрузку.  
Здесь такой же случай как и с get'ом. "ОК" - дефолтный ответ, который приведен в документации https://redis.io/commands/set/

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/RedisLikeSpringAppApplication.java:
```diff
@@ -0,0 +1,19 @@
+package com.github.theCoolerSuptelov.redisLikeInMemory.redisLikeSpringApp;
```
в именах пакетов заглавные буквы не используем. максимум подчеркивания для раздленеия слов, но это если слова больше двух, а такого лучше не допускать.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+
+@Service
+public class CacheImplService {
+    public static final String OK = "OK";
```
что такое ок? сообщение? ответ? код? почему константа публична?

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+        }
+        return deletionCounter;
+    }
+    @Secured("ROLE_ADMIN")
```
Зачем это защищать еще раз, если важнее всего делать ткую защиту именно в контроллере?

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+
+    @Value("${defaultSavePathCacheImpl}")
+    private String savePath;
+    private final long defaultTtlMillis = 60_000;
```
это лучше вынести в контсанты.

AndryLat replied:
то есть добавить статический контекст и именовать как онстанты (заглавными буквами)

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+    private String savePath;
+    private final long defaultTtlMillis = 60_000;
+
+    private HashMap<String, CacheUnit> cacheHM = new HashMap<>();
```
чт отакое HM в имени? зачем указывать тип переменной в ее имени? а если завтра тим поменяет на три мэп переименовывать по всему коду?

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+
+    private HashMap<String, CacheUnit> cacheHM = new HashMap<>();
+
+    public Integer deleteKeys(List<String> arrKeysToDelete){
```
имени keys будет с головой достаточно.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+                                  Boolean setKeyIfNotExist,
+                                  Boolean setKeyIfExist,
+                                  Boolean keepCurrentTtl){
+
```
я бы добавил валидации на значения.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+            return null;
+        }
+        if (setKeyIfNotExist && cacheUnit != null){
+            return null;
```
возврат нала не очень хорошо. лучше Optional/

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+            return OK;
+        }
+
+        CacheUnit freshCacheUnit = new CacheUnit(defineExpireTimeMilliseconds(exSeconds, pxMilliseconds,unixTimeExpireSeconds,unixTimeExpireMilliseconds)
```
так форматировать не хорошо. вынеси результат метода в переменную, а тут в конструктор передай переменную. будет читабельнее.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+        return deletionCounter;
+    }
+    @Secured("ROLE_ADMIN")
+    public String putCachingValue(CachedValueDTO cachedValueDTO,
```
Вместо такой кучи переменных лучше сделать объект и его передавать. А чтобы было проще на контроллере - принимай джейсон.

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+            ObjectInputStream objectInputStream
+                    = new ObjectInputStream(new FileInputStream(this.getSavePath()));
+            cacheHM = (HashMap<String, CacheUnit>) objectInputStream.readObject();
+        } catch (IOException e) {
```
одинаковую обработку можно заменить оператором |

AndryLat commented on src/main/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/Services/CacheImplService.java:
```diff
@@ -0,0 +1,154 @@
+public class CacheImplService {
+    public static final String OK = "OK";
+
+    public String getSavePath() {
```
это выглядит как геттер поля класса, не надо так. и также не понимаю, почему геттер один, сеттер другой, то есть когда засейвим, мы не сможем взять с этого другого места?

AndryLat commented on src/test/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/RedisLikeSpringAppApplicationTests.java:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	@Test
+	public void defaultRestApiPipeline() throws Exception {
```
а) используй нейминг https://dzone.com/articles/7-popular-unit-test-naming
б) один громадный тест, больше похожий на интеграционный. каждая операция должна быть протестирована отдельно.

AndryLat commented on src/test/java/com/github/theCoolerSuptelov/redisLikeInMemory/resources/RedisLikeApp.json:
```diff
@@ -0,0 +1,893 @@
+							"body": "exercitation nostrud"
+						}
+					]
+				},
```
что это и зачем оно все тут?

AndryLat commented on src/main/resources/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM openjdk:11
```
этот файл должен быть в корне

AndryLat commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,6 @@
+defaultSavePathCacheImpl= D:/java/petProjects/redisLikeSpringApp/storage/RedisLikeApp.txt
```
полный путь - всегда отличная идея, особенно, кроссплатформенная:)

AndryLat commented on src/test/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/RedisLikeSpringAppApplicationTests.java:
```diff
@@ -0,0 +1,135 @@
+
+@SpringBootTest
+@AutoConfigureMockMvc
+class RedisLikeSpringAppApplicationTests {
```
SpringAppApplication ? а под App не Application подразумевается?

AndryLat commented:
в общем, есть проблемы с арзитектурой приложения.
а) ответы всегда надо формировать явно и формировать качественные ответы. хотя бы джейсон с кодом и полем для ошибки есть такая происходит.
б) не хватает обработки ситуаций, когда что-то пойдет не так.
в) подход к удалению записей, они удаляются, когда их еще раз запросили, а что если к ним больше никогда не обратятся? так и будут забивать кэш.

TheCoolerSuptelov committed:
changes after review

TheCoolerSuptelov committed:
renamed:    src/main/resources/Dockerfile -> src/Dockerfile
	modified:   src/Dockerfile
	modified:   src/test/java/com/github/theCoolerSuptelov/redisLikeInMemory/redisLikeSpringApp/RedisLikeSpringAppApplicationTests.java

TheCoolerSuptelov committed:
renamed:    src/Dockerfile -> Dockerfile
	modified:   Dockerfile
